### PR TITLE
replace boost:bind() with lambdas

### PR DESCRIPTION
**Pull request by neheb**
lambdas are standard C++. bind is also recommended against by LLVM.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

### Checklist

I have:
- [ ] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
Thank you for this merge request. It looks like it doesn't compile, would you mind having a look?

**Comment by neheb** [comment]
Partially why I made it a draft :).

I think the proper way forward is to merge the other PR and to rebase this one.

**Comment by neheb** [comment]
Looks like it builds. No idea why it's failing.

File: pdns/pdns_recursor.cc [hunk 1]
```diff
@@ -2846,7 +2846,7 @@ static void handleNewUDPQuestion(int fd, FDMultiplexer::funcparam_t& var)
           }
 
           if(g_weDistributeQueries) {
-            distributeAsyncFunction(data, boost::bind(doProcessUDPQuestion, data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues));
+            distributeAsyncFunction(data, [&]{ return doProcessUDPQuestion(data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues); });
```
Comment (rgacogne) [inline pdns/pdns_recursor.cc:2849]: This version is capturing a `thread_local` data by reference and passing it to a function that will get executed by a different thread, and also passing `proxyProtocolValues` by reference to that same thread while the initial thread might be altering it at the same time. 
```suggestion
            distributeAsyncFunction(data, [data = data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues]() mutable { return doProcessUDPQuestion(data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues); });
```

Perhaps we should stop using a `thread_local` data here, as that optimization still results in a copy when we are distributing anyway.

File: pdns/rec_channel_rec.cc [hunk 2]
```diff
@@ -689,9 +689,9 @@ static string doClearTA(T begin, T end)
     g_luaconfs.modify([entry](LuaConfigItems& lci) {
         lci.dsAnchors.erase(entry);
       });
-    broadcastAccFunction<uint64_t>(boost::bind(pleaseWipeCache, entry, true, 0xffff));
-    broadcastAccFunction<uint64_t>(boost::bind(pleaseWipePacketCache, entry, true, 0xffff));
-    broadcastAccFunction<uint64_t>(boost::bind(pleaseWipeAndCountNegCache, entry, true));
+    broadcastAccFunction<uint64_t>([&]{return pleaseWipeCache(entry, true, 0xffff);});
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc:692]:
```suggestion
    broadcastAccFunction<uint64_t>([=]{return pleaseWipeCache(entry, true, 0xffff);});
```

File: pdns/rec_channel_rec.cc [hunk 3]
```diff
@@ -689,9 +689,9 @@ static string doClearTA(T begin, T end)
...
+    broadcastAccFunction<uint64_t>([&]{return pleaseWipePacketCache(entry, true, 0xffff);});
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc:693]:
```suggestion
    broadcastAccFunction<uint64_t>([=]{return pleaseWipePacketCache(entry, true, 0xffff);});
```

File: pdns/rec_channel_rec.cc [hunk 4]
```diff
@@ -689,9 +689,9 @@ static string doClearTA(T begin, T end)
...
+    broadcastAccFunction<uint64_t>([&]{return pleaseWipeAndCountNegCache(entry, true);});
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc:694]:
```suggestion
    broadcastAccFunction<uint64_t>([=]{return pleaseWipeAndCountNegCache(entry, true);});
```

File: pdns/pdns_recursor.cc [hunk 5]
```diff
@@ -3516,7 +3516,7 @@ template<class T> T broadcastAccFunction(const boost::function<T*()>& func)
 
     const auto& tps = threadInfo.pipes;
     ThreadMSG* tmsg = new ThreadMSG();
-    tmsg->func = boost::bind(voider<T>, func);
+    tmsg->func = [&]{ return voider<T>(func); };
```
Comment (rgacogne) [inline pdns/pdns_recursor.cc:3519]:
```suggestion
    tmsg->func = [func]{ return voider<T>(func); };
```

**Comment by neheb** [comment]
Done. Build passes.

**Comment by neheb** [comment]
Marked as ready.

File: pdns/pdns_recursor.cc [hunk 6]
```diff
@@ -2846,7 +2846,8 @@ static void handleNewUDPQuestion(int fd, FDMultiplexer::funcparam_t& var)
           }
 
           if(g_weDistributeQueries) {
-            distributeAsyncFunction(data, boost::bind(doProcessUDPQuestion, data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues));
+            distributeAsyncFunction(data, [data = data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues]() mutable
```
Comment (rgacogne) [inline pdns/pdns_recursor.cc:2849]: While the other changes are in places where performance does not really matter much, this one is absolutely critical. Perhaps we should benchmark that change before merging this PR? @omoerbeek ?

**Comment by omoerbeek** [comment]
An somewhat unscientific test:
```
#include <boost/function.hpp>
#include "iputils.hh"
#include "proxy-protocol.hh"

typedef boost::function<void*(void)> pipefunc_t;

void distributeAsyncFunction(const std::string& packet, const pipefunc_t& func)
{
  
}

static string* doProcessUDPQuestion(const std::string& question, const ComboAddress& fromaddr, const ComboAddress& destaddr, ComboAddress source, ComboAddress destination, struct timeval tv, int fd, std::vector<ProxyProtocolValue>& proxyProtocolValues)
{
  return nullptr;
}

int main(int argc, char *argv[])
{
  static thread_local std::string data;
  ComboAddress fromaddr;
  ComboAddress source;
  ComboAddress destination, dest;
  struct timeval tv={0,0};
  int fd;
  std::vector<ProxyProtocolValue> proxyProtocolValues;

  int count = 10000000;
  if (strcmp(argv[1], "bind") == 0) {
    cerr << "bind" << endl;
    for (int i = 0; i < count; i++) {
      distributeAsyncFunction(data, boost::bind(doProcessUDPQuestion, data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues));
    }
  } else {
    cerr << "lambda" << endl;
    for (int i = 0; i < count; i++) {
      distributeAsyncFunction(data, [data = data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues]() mutable
                                    { return doProcessUDPQuestion(data, fromaddr, dest, source, destination, tv, fd, proxyProtocolValues); });
    }
  }
}
```

On my OpenBSD x1 Carbon laptop
```
[otto@x1:356]$ c++ --version
OpenBSD clang version 10.0.1 
Target: amd64-unknown-openbsd6.7
Thread model: posix
InstalledDir: /usr/bin
[otto@x1:357]$ c++ -I. -I /usr/local/include -O2 x.cc 
[otto@x1:358]$  time ./a.out bind
bind
    0m05.58s real     0m05.54s user     0m00.01s system
[otto@x1:359]$ time ./a.out x
lambda
    0m04.58s real     0m04.55s user     0m00.00s system
```

And a CentOS8 VM:
```
otto@otto:~/pdns/pdns/recursordist$ c++ --version
c++ (Debian 8.3.0-6) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

otto@otto:~/pdns/pdns/recursordist$ time ./a.out bind
bind

real    0m2.825s
user    0m2.823s
sys     0m0.000s
otto@otto:~/pdns/pdns/recursordist$ time ./a.out x
lambda

real    0m1.820s
user    0m1.820s
sys     0m0.000s
```

**Comment by rgacogne** [comment]
So.. ship it? :-)

**Comment by Habbie** [comment]
NICE

**Review by omoerbeek — APPROVED** [review]

**Pull request merged**